velox/functions/prestosql/DateTimeImpl.h

```diff
@@ -47,11 +47,11 @@ FOLLY_ALWAYS_INLINE std::optional<Timestamp> fromUnixtime(double unixtime) {
   static const Timestamp kMinTimestamp(
       kMin / 1000 - 1, (kMin % 1000 + 1000) * kNanosecondsInMillisecond);
 
-  if (UNLIKELY(unixtime >= kMax)) {
+  if (UNLIKELY(unixtime >= static_cast<double>(kMax))) {
```

Comment:
I'm not sure what Presto's implementation does, but wouldn't it make more sense to compare static_cast<int64_t>(floor(unixtime)) to kMax/kMin since seconds is floor(unixtime) and gets implicitly cast to int64_t when passed to Timestamp (which is where the risk of over/underflow is)

and casting from int64_t to double can lose some precision, though I'm not sure if it does for those two particular numbers, but if it does, unixtime could end up being larger than kMax but still fail this check

Reply:
Not sure if floor(unixtime) is right, since technically range of double is 10 ^ -304 to 10 ^ 308 which is more than what is represented by int64. Converting kMax to double will land in this range and allows us to know whether passed value is > max/min. This is because i dont think casting floor(unixtime) to int is reliable (or maybe right). 

@majetideepak can you comment ?

Reply:
@kevinwilfong Here is Presto's implementation:

```
    @ScalarFunction("from_unixtime")
    @SqlType(StandardTypes.TIMESTAMP)
    public static long fromUnixTime(@SqlType(StandardTypes.DOUBLE) double unixTime)
    {
        return Math.round(unixTime * 1000);
    }
```

Reply:
I realize what I said above doesn't make sense because of the limits of int64_t.

If you look below at line 62, we're using floor(unixtime) to compute seconds, and in line 64 seconds gets converted to int64_t implicitly by passing  it to the Timestamp constructor.

I'm trying to avoid this case where if you cast kMax to a double you actually get a number larger than kMax
https://stackoverflow.com/questions/65749055/when-a-64bit-int-is-cast-to-64bit-float-in-c-c-and-doesnt-have-an-exact-match

so if unixtime is a fractional value in between kMax and static_cast<double>(kMax) you'll miss this case

another way to do this might be to cast kMax to a uint64_t add 1 to it, and then cast it to a double (I think this can be represented exactly as a double), for kMin I think the way it is is fine because I think that can be represented as a double

Reply:
Yeah `if ((1ull << 63) - unixTime >= 1)` would probably be the safest

Reply:
I made this change.